Added unnecessary-else Rule

debsmita1 opened the pull request:
#### PR checklist

- [x] Addresses an existing issue: fixes #4429
- [x] New feature
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
This Rule disallows `else` blocks following `if` blocks ending with a `return` , `throw`, `break` or `continue`statement.


#### Is there anything you'd like reviewers to focus on?
This is my very first open-source contribution, so all kinds of feedback is appreciated.


#### CHANGELOG.md entry:
[new-rule] unnecessary-else﻿

palantirtech commented:
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @debsmita1! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

JoshuaKGoldberg commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
```
Hmm, this rationale seems to restate the rule description. How about something like:

> ```
> When an `if` block is guaranteed to exit control flow when entered,
> it is unnecessary to add an `else` statement.
> The contents that would be in the `else` block can be placed after the end of the `if` block.
> ```

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/a1ec00c3360cd8d8b7e36e4ea7b18410a14db156

JoshuaKGoldberg commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
```
TSLint's trying to move to more descriptive documentation with code samples for rules. Please add a `codeExamples` section to this rule.

You can see rules such as `arrowReturnShorthandRule.ts` for examples of existing ones.

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/b46d5385f6d5848136581f593a1a465db4990ffe

JoshuaKGoldberg commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+    /* tslint:disable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string): string {
+        return `If block contains \`${name}\` statement. Consider replacing the contents of else block outside of the block.`;
```
The rule doesn't need to say "consider" because the rule failure is a definite "you must do this":

```suggestion
        return `The preceding \`if\` block ends with a \`${name}\` statement. This \`else\` block is unnecessary.`;
```

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/a1ec00c3360cd8d8b7e36e4ea7b18410a14db156

JoshuaKGoldberg commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+    | ts.ReturnStatement;
+
+function walk(ctx: Lint.WalkContext<void>): void {
+    let inElse = false;
```
It looks like `walk` will consider any jump statement within an `if` block to make that block not need an `else`. What if there's one within a nested `if` statement?

```ts
if (something) {
    if (anotherThing) {
        return;
    } else {
        // Unnecessary else block
    }
} else {
    // Reasonable else block
}
```

Instead of walking down within the `if`, I think you can get away with just checking whether its _last_ statement is a jump statement. In other words, I think you should be able to just use `if (isIfStatement(node)) {` and check the last statement within the `thenStatement` and the node's `elseStatement`.

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/89c814eb7648cbe716af317adb05add694305430

JoshuaKGoldberg commented on test/rules/no-unnecessary-else/test.ts.lint:
```diff
@@ -0,0 +1,150 @@
+}
+
+const testBreak = () => {
+    switch () {
```
It's invalid to have empty contents in `case`, `if`, and `switch` statements. Rules can behave unpredictably when there are syntax errors, so tests should have valid syntax whenever possible.

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/89c814eb7648cbe716af317adb05add694305430

JoshuaKGoldberg commented on test/rules/no-unnecessary-else/test.ts.lint:
```diff
@@ -0,0 +1,150 @@
+        if () {
+            return ;
+        } else {
+            return ;
```
Here's an example of the rule not giving a complaint when it should. Unclear if it's from the syntax errors in this test file or from the structural notes above.

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/89c814eb7648cbe716af317adb05add694305430

debsmita1 commented:
const testReturn = (a) => {
      if (a%2 === 0) {
         return "multiple of 2" ;
      } else if (a%3 === 0) {
          return "multiple of 3" ;
      } else if (a%5 === 0) {
          return "multiple of 5" ;
      } else {
          return "not a multiple of 2 , 3 or 5" ;
      }
  }
For the above scenario what should be the correct behavior ? The linter must throw an error at every `else` block ? @JoshuaKGoldberg

debsmita1 commented on src/configs/recommended.ts:
```diff
@@ -91,6 +91,7 @@ export const rules = {
     "no-string-throw": true,
     "no-switch-case-fall-through": false,
     "no-trailing-whitespace": true,
+    "no-unnecessary-else": true,
```
Done https://github.com/palantir/tslint/pull/4502/commits/8d4f035a89b74f342aa687724425c8b5406d077d

debsmita1 commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+            When control flow statements,
+            such as \`return\`, \`continue\`, \`break\` or \`throw\` is written inside \`if\` block,
+            Then \`else\` block becomes unnecessary.`,
+        ruleName: "no-unnecessary-else",
```
Done https://github.com/palantir/tslint/pull/4502/commits/8d4f035a89b74f342aa687724425c8b5406d077d

debsmita1 commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+/**
+ * @license
+ * Copyright 2018 Palantir Technologies, Inc.
```
Done https://github.com/palantir/tslint/pull/4502/commits/ee45076c830524d83245b806e4a65bba5a2ca5da

debsmita1 commented on test/rules/no-unnecessary-else/test.ts.lint:
```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+//valid
```
Done https://github.com/palantir/tslint/pull/4502/commits/c2db576ebe200c4e74417e56954613d16a3292b9

debsmita1 commented on src/rules/noUnnecessaryElseRule.ts:
```diff
@@ -0,0 +1,98 @@
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        description: Lint.Utils.dedent`
+            Disallows else block when the If block contains control flow statements, such as \`return\`, \`continue\`,
```
Done https://github.com/palantir/tslint/pull/4502/commits/a1ec00c3360cd8d8b7e36e4ea7b18410a14db156

debsmita1 commented:
I have addressed all your comments.
![screenshot from 2019-02-17 18-42-01](https://user-images.githubusercontent.com/22490998/52913391-c5dbeb80-32e3-11e9-8456-fe484b8c964f.png)
Cant figure out why these 2 are failing. Could you please help @JoshuaKGoldberg

JoshuaKGoldberg commented on src/rules/unnecessaryElseRule.ts:
```diff
@@ -0,0 +1,100 @@
+    ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (utils.isIfStatement(node)) {
+            const jumpStatement = getJumpStatement(
+                getLastStatement(node.thenStatement as ts.Block),
```
Fun fact: `node.thenStatement` could also be an `ExpressionStatement`:

```ts
if (condition) someExpression;
```

...because of the lack of `{}`s. You'll need to handle that & have test cases for it as well.

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/be2f8b4323b7b534130489a8467a91b6962d745e

JoshuaKGoldberg commented on src/rules/unnecessaryElseRule.ts:
```diff
@@ -0,0 +1,100 @@
+            }
+            ts.forEachChild(node.thenStatement, cb);
+        } else {
+            return ts.forEachChild(node, cb);
```
It looks like you have a few different `ts.forEachChild`s: individual ones for the node's `elseStatement` and `thenStatement` if it's an `IfStatement`, but just a regular recursion otherwise? You should be able to simplify this a bit by only having one call to `ts.forEachChild` that's always called.

Also, if `jumpStatement` is undefined but `node.elseStatement` isn't, this code will skip it. That seems like a potential bug?

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/be2f8b4323b7b534130489a8467a91b6962d745e

JoshuaKGoldberg replied:
Sorry @debsmita1 I think I wasn't very clear here. I'm suggesting you don't need to have `ts.forEachChild` statements in the big `if (utils.isIfStatement` block:

```ts
if (utils.isIfStatement(node)) {
    // all your glorious rule logic here
    // no ts.forEachChild in this area
}

ts.forEachChild(node, cb);
```

Note the lack of `else` statement: a single `ts.forEachChild` should recurse on all the node's children no matter what. Could you please switch to that? It's a bit less code and better practice to minimize recursive calls (since they can get tricky in larger rules).

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/f36d4201e2fa8ec3aec60a022db2bf51435aceaf

JoshuaKGoldberg commented on src/rules/code-examples/unnecessaryElse.examples.ts:
```diff
@@ -0,0 +1,76 @@
+            "rules": { "unnecessary-else": true }
+        `,
+        pass: Lint.Utils.dedent`
+            if () {
```
Coming back to these after a bit, the blank `()s` could be a little confusing for newcomers because they're invalid syntax in our code examples. Could you change it to `if (someCondition())` and similar?

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/a2842a9ad5ba17d0bb597296ad99b7bc0bbe5142

JoshuaKGoldberg commented on src/rules/unnecessaryElseRule.ts:
```diff
@@ -0,0 +1,117 @@
+    /* tslint:disable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string): string {
+        return `The preceding \`if\` block ends with a \`${name}\` statement. This \`else\` block is unnecessary.`;
```
Sorry to nitpick on this again: I think it might be a little more clear if the error message said <code>... This \`else\` is unnecessary</code> _(removing the `block`)_. Technically, the block's contents are fine; it's just the `else` that's unnecessary. Would you mind changing it?

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/f36d4201e2fa8ec3aec60a022db2bf51435aceaf

JoshuaKGoldberg commented on src/rules/unnecessaryElseRule.ts:
```diff
@@ -0,0 +1,117 @@
+            let jumpStatement;
+            if (
+                node.thenStatement.kind !== ts.SyntaxKind.Block &&
+                !utils.isSameLine(ctx.sourceFile, node.thenStatement.pos, node.thenStatement.end)
```
I'm a little confused, is it actually necessary to check `utils.isSameLine`? `node` and its `.thenStatement` should behave the same regardless of whitespace position, right?

debsmita1 replied:
I have verified this multiple times,
if() return;
else return;

behaves very differently from 

if()
	return;
else
	return;

which is why I had to check `utils.isSameLine`. Could you kindly suggest a way to optimize this ? @JoshuaKGoldberg

JoshuaKGoldberg replied:
Interesting! In what ways are you finding them to be different?

With this test, I'm getting the same result reported for both:

```ts
function testSameLine(condition) {
    if (condition) return 'a';
    else return 'b';
}

function testNextLine(condition) {
    if (condition)
    return 'a';
    else
    return 'b';
}

for (const condition of [true, false]) {
    console.log(condition);
    console.log("\tSame line:", testSameLine(condition));
    console.log("\tNext line:", testNextLine(condition));
    console.log("");
}
```

```
true
    Same line: a
    Same line: a

false
    Same line: b
    Same line: b
```

debsmita1 replied:
If I remove any of the `utils.isSameLine` checks I get the below error and the rule doesnt work
![logs](https://user-images.githubusercontent.com/22490998/53304449-50e85300-389b-11e9-9b8c-834727767a51.png)

JoshuaKGoldberg replied:
Referring to the `TypeError; cannot read property '0' of undefined`?

It looks like the problem is from the `jumpStatement = getJumpStatement(getLastStatement(node.thenStatement as ts.Block));`. The rule can't assume `node.thenStatement as ts.Block` is valid because `node.thenStatement` might be some other kind of node, such as a `ReturnStatement` in the sample code in this thread.

You should be able to simplify the cases within the `if (utils.isIfStatement` block into just the following:

* `tsutils.isBlock(node.thenStatement)`
* Not ☝

...rather than the three cases you have now.

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/f36d4201e2fa8ec3aec60a022db2bf51435aceaf

debsmita1 replied:
Done https://github.com/palantir/tslint/pull/4502/commits/4c29b4309e698117979d5bd9a8e62b9fbc337b1f

JoshuaKGoldberg commented:
> ![screenshot from 2019-02-17 18-42-01](https://user-images.githubusercontent.com/22490998/52913391-c5dbeb80-32e3-11e9-8456-fe484b8c964f.png)

I also can't figure out why those are failing at first glance. Very spooky. If they're still failing after the introduced lint complaints are fixed, we can take a deeper look...

adidahiya commented on src/rules/unnecessaryElseRule.ts:
```diff
@@ -0,0 +1,100 @@
+        it is unnecessary to add an \`else\` statement.
+        The contents that would be in the \`else\` block can be placed after the end of the \`if\` block.`,
+        ruleName: "unnecessary-else",
+        type: "functionality",
```
I think this is a purely stylistic rule; is there any functional difference?
